nntrainer/graph/graph_core.h

```diff
@@ -224,14 +241,37 @@ class GraphCore {
    * @details   Ensures that the node has a unique and a valid name. A valid
    * name pre-assigned to the node can be changed if force_rename is enabled.
    */
-  void ensureName(std::shared_ptr<GraphNode> &node,
-                  const std::string &prefix = "",
+  void ensureName(GraphNode &node, const std::string &prefix = "",
                   const std::string &postfix = "", bool force_rename = false);
 
+  void remove_last_node() {
+    auto last_node = Sorted.back();
+    Sorted.pop_back();
+    adj.erase(adj.begin() + last_node->getIndex());
+
+    /**
+     * Remove all the connections for the current lasy layer as it will now only
+     */
+    last_node = Sorted.back();
+    adj[last_node->getIndex()].resize(1);
+  }
+
+  void addLossToSorted() { Sorted.push_back(adj.back().front()); }
+
+  bool verifyNode(const std::string &name) {
+    if (node_names.find(name) == node_names.end())
+      return false;
+    return true;
+  }
+
 private:
   std::vector<std::list<std::shared_ptr<GraphNode>>>
     adj; /**< adjacency list for graph */
+  std::vector<std::shared_ptr<GraphNode>> node_list; /**< Ordered Node List  */
+
   std::vector<std::shared_ptr<GraphNode>> Sorted; /**< Ordered Node List  */
+  bool sorted; /** if the node_list is sorted */
+
   std::set<std::string>
```

Comment:
Not related to this commit but we could substitute this to unordered_set?

Reply:
added a TODO.

Reply:
Alright